File: apps/api/app/models/event.py
"""Event model."""

import enum
import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import DateTime, ForeignKey, Text
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class EventType(enum.StrEnum):
    SWARM = "swarm"
    SPLIT = "split"
    COMBINE = "combine"
    REQUEEN = "requeen"
    FEED = "feed"
    WINTER_PREP = "winter_prep"


class Event(Base):
    __tablename__ = "events"

    hive_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("hives.id", ondelete="CASCADE"), nullable=False, index=True
    )
    event_type: Mapped[EventType] = mapped_column(
        SAEnum(EventType, name="event_type"), nullable=False
    )
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    details: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relationships
    hive: Mapped["Hive"] = relationship("Hive", back_populates="events")

File: apps/mobile/database/models/Event.ts
import { Model } from "@nozbe/watermelondb";
import { text, field, date, relation, readonly } from "@nozbe/watermelondb/decorators";
import type { Associations } from "@nozbe/watermelondb/Model";

export default class Event extends Model {
  static table = "events";

  static associations: Associations = {
    hives: { type: "belongs_to" as const, key: "hive_id" },
  };

  @text("hive_id") hiveId!: string;
  @text("event_type") eventType!: string;
  @date("occurred_at") occurredAt!: Date;
  @text("details_json") detailsJson!: string | null;
  @text("notes") notes!: string | null;
  @readonly @date("created_at") createdAt!: Date;
  @readonly @date("updated_at") updatedAt!: Date;

  @relation("hives", "hive_id") hive!: any;

  get details(): Record<string, unknown> | null {
    if (!this.detailsJson) return null;
    try { return JSON.parse(this.detailsJson); } catch { return null; }
  }
}
